gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -0,0 +1,828 @@
+      TableMetadata tableMetadata, TableIdentifier tid) throws IOException {
+    PartitionSpec partitionSpec = table.spec();
+    Transaction transaction = tableMetadata.transaction.get();
+    Set<DataFile> oldDataFiles =
+        getIcebergDataFilesToBeDeleted(gmce, table, new HashMap<>(), oldSpecsMap, partitionSpec);
+    tableMetadata.deleteFiles = Optional.of(tableMetadata.deleteFiles.or(transaction::newDelete));
+    DeleteFiles deleteFiles = tableMetadata.deleteFiles.get();
+    oldDataFiles.stream().forEach(deleteFiles::deleteFile);
+    parallelRunner.submitCallable(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        try {
+          long olderThan = getExpireSnapshotTime();
+          long start = System.currentTimeMillis();
+          ExpireSnapshots expireSnapshots = table.expireSnapshots();
+          final Table tmpTable = table;
+          expireSnapshots.deleteWith(new Consumer<String>() {
+            @Override
+            public void accept(String file) {
+              if (file.startsWith(tmpTable.location())) {
+                tmpTable.io().deleteFile(file);
+              }
+            }
+          }).expireOlderThan(olderThan).commit();
+          //TODO: emit these metrics to Ingraphs, in addition to metrics for publishing new snapshots and other Iceberg metadata operations.
+          log.info("Spent {} ms to expire snapshots older than {} ({}) in table {}", System.currentTimeMillis() - start,
+              new DateTime(olderThan).toString(), olderThan, tid.toString());
+        } catch (Exception e) {
+          log.error(String.format("Fail to expire snapshots for table %s due to exception ", tid.toString()), e);
+        }
+        return null;
+      }
+    }, tid.toString());
+  }
+
+  private long getExpireSnapshotTime() {
+    PeriodFormatter periodFormatter = new PeriodFormatterBuilder().appendYears()
+        .appendSuffix("y")
+        .appendMonths()
+        .appendSuffix("M")
+        .appendDays()
+        .appendSuffix("d")
+        .appendHours()
+        .appendSuffix("h")
+        .appendMinutes()
+        .appendSuffix("m")
+        .toFormatter();
+    return DateTime.now()
+        .minus(periodFormatter.parsePeriod(
+            conf.get(EXPIRE_SNAPSHOTS_LOOKBACK_TIME, DEFAULT_EXPIRE_SNAPSHOTS_LOOKBACK_TIME)))
+        .getMillis();
+  }
+
+  protected void addFiles(GobblinMetadataChangeEvent gmce, Map<String, Collection<HiveSpec>> newSpecsMap, Table table,
+      TableMetadata tableMetadata) throws IOException {
+    PartitionSpec partitionSpec = table.spec();
+    Transaction transaction = tableMetadata.transaction.get();
+    tableMetadata.appendFiles = Optional.of(tableMetadata.appendFiles.or(transaction::newAppend));
+    AppendFiles appendFiles = tableMetadata.appendFiles.get();
+    getIcebergDataFilesToBeAdded(gmce.getNewFiles(), partitionSpec, newSpecsMap,
```

Comment:
Do you think this long statement can be separated out as a private helper method as rewrite is also using ?